File: backend/app/services/modem_manager.py
"""
ModemManager integration via mmcli CLI tool.
Supports multiple USB 4G modems simultaneously.
"""
import subprocess
import json
import re
import asyncio
import logging
from typing import List, Optional, Dict, Any
from datetime import datetime

logger = logging.getLogger(__name__)


def _run(cmd: List[str], timeout: int = 30) -> tuple[int, str, str]:
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
    return result.returncode, result.stdout.strip(), result.stderr.strip()


def list_modems() -> List[Dict[str, Any]]:
    """Return all modems detected by ModemManager."""
    code, out, err = _run(["mmcli", "-L", "-J"])
    if code != 0:
        logger.error(f"mmcli -L failed: {err}")
        return []
    try:
        data = json.loads(out)
        paths = data.get("modem-list", [])
        modems = []
        for path in paths:
            info = get_modem_info(path)
            if info:
                modems.append(info)
        return modems
    except (json.JSONDecodeError, KeyError) as e:
        logger.error(f"Failed to parse modem list: {e}")
        return []


def get_modem_info(mm_path: str) -> Optional[Dict[str, Any]]:
    """Get detailed info for a single modem by its D-Bus object path."""
    # Extract modem index from path like /org/freedesktop/ModemManager1/Modem/0
    match = re.search(r"/Modem/(\d+)$", mm_path)
    if not match:
        return None
    idx = match.group(1)
    code, out, err = _run(["mmcli", "-m", idx, "-J"])
    if code != 0:
        logger.error(f"mmcli -m {idx} failed: {err}")
        return None
    try:
        data = json.loads(out)
        m = data["modem"]
        generic = m.get("generic", {})
        threegpp = m.get("3gpp", {})
        access_techs = generic.get("access-technologies", [])
        if isinstance(access_techs, list):
            access_technologies = ",".join(access_techs)
        else:
            access_technologies = str(access_techs)

        reg_state = threegpp.get("registration-state", "") or ""

        bearer_stats = _get_bearer_stats(idx)
        sim_info = _get_sim_info(generic.get("sim", ""))
        bands = generic.get("current-bands", [])
        current_bands = ",".join(bands) if isinstance(bands, list) else str(bands)
        ports = generic.get("ports", [])
        ports_str = ",".join(ports) if isinstance(ports, list) else str(ports)
        return {
            "mm_object_path": mm_path,
            "mm_index": idx,
            "device_path": generic.get("primary-port", ""),
            "manufacturer": generic.get("manufacturer", ""),
            "model": generic.get("model", ""),
            "imei": threegpp.get("imei", ""),
            "operator": threegpp.get("operator-name", ""),
            "signal_quality": int(generic.get("signal-quality", {}).get("value", 0)),
            "raw_state": generic.get("state", "unknown"),
            "status": _map_state(generic.get("state", "unknown")),
            "phone_number": _parse_own_number(generic) or _get_phone_number(idx),
            "access_technologies": access_technologies,
            "registration_state": reg_state,
            "tx_bytes": bearer_stats.get("tx_bytes", 0),
            "rx_bytes": bearer_stats.get("rx_bytes", 0),
            "connection_duration": bearer_stats.get("connection_duration", 0),
            "imsi": sim_info.get("imsi", ""),
            "iccid": sim_info.get("iccid", ""),
            "firmware_revision": generic.get("revision", ""),
            "hardware_revision": generic.get("hardware-revision", ""),
            "current_bands": current_bands,
            "sim_operator_name": sim_info.get("sim_operator_name", ""),
            "sim_operator_code": sim_info.get("sim_operator_code", ""),
            "current_modes": generic.get("current-modes", ""),
            "ports": ports_str,
            "plugin": generic.get("plugin", ""),
        }
    except (json.JSONDecodeError, KeyError) as e:
        logger.error(f"Failed to parse modem info: {e}")
        return None


def _get_bearer_stats(idx: str) -> dict:
    """Fetch TX/RX bytes and connection duration from the active bearer."""
    code, out, _ = _run(["mmcli", "-m", idx, "--list-bearers", "-J"])
    if code != 0:
        return {}
    try:
        data = json.loads(out)
        paths = data.get("modem.bearers.dbus-path", []) or data.get("bearer-list", [])
        if not paths:
            return {}
        bearer_path = paths[0]
        match = re.search(r"/Bearer/(\d+)$", bearer_path)
        if not match:
            return {}
        b_idx = match.group(1)
        code2, out2, _ = _run(["mmcli", "-b", b_idx, "-J"])
        if code2 != 0:
            return {}
        b = json.loads(out2).get("bearer", {})
        stats = b.get("stats", {})
        status = b.get("status", {})
        return {
            "tx_bytes": int(stats.get("tx-bytes", 0) or 0),
            "rx_bytes": int(stats.get("rx-bytes", 0) or 0),
            "connection_duration": int(status.get("connection-duration", 0) or 0),
        }
    except Exception:
        return {}


def _get_sim_info(sim_path: str) -> dict:
    """Fetch IMSI and ICCID from the SIM object."""
    if not sim_path:
        return {}
    match = re.search(r"/SIM/(\d+)$", sim_path)
    if not match:
        return {}
    sim_idx = match.group(1)
    code, out, _ = _run(["mmcli", "-i", sim_idx, "-J"])
    if code != 0:
        return {}
    try:
        props = json.loads(out).get("sim", {}).get("properties", {})
        return {
            "imsi": props.get("imsi", ""),
            "iccid": props.get("iccid", ""),
            "sim_operator_name": props.get("operator-name", ""),
            "sim_operator_code": props.get("operator-code", ""),
        }
    except Exception:
        return {}


def _parse_own_number(generic: dict) -> str:
    numbers = generic.get("own-numbers", [])
    if isinstance(numbers, list) and numbers:
        return numbers[0]
    return ""


def _get_phone_number(idx: str) -> str:
    # AT+CNUM 通过 mmcli 透传 AT 命令给调制解调器，部分 SIM 卡不支持此命令会返回空
    code, out, err = _run(["mmcli", "-m", idx, "--command=AT+CNUM", "-J"])
    if code != 0:
        return ""
    try:
        data = json.loads(out)
        response = data.get("modem", {}).get("command", {}).get("response", "")
        match = re.search(r'\+CNUM:.*?"(\+?\d+)"', response)
        return match.group(1) if match else ""
    except Exception:
        return ""


def _map_state(state: str) -> str:
    mapping = {
        "registered": "connected",
        "connected": "connected",
        "disabled": "disconnected",
        "disabling": "disconnected",
        "enabling": "disconnected",
        "searching": "disconnected",
        "failed": "error",
    }
    return mapping.get(state.lower(), "unknown")


def send_sms(mm_index: str, phone_number: str, text: str) -> tuple[bool, str]:
    """Send an SMS via a specific modem. Returns (success, message).

    mmcli 短信发送是三步对象操作（非单条命令）：
    1. --messaging-create-sms  → 创建 SMS 对象，得到 /SMS/<n> 路径
    2. -s <n> --send            → 实际发送
    3. -s <n> --delete          → 清理已发送对象（否则占用设备内存）

    --messaging-create-sms 的值按逗号解析，text 内若含逗号会被截断，
    故用 text="..." 引号形式将内容作为整体传入。
    """
    escaped = text.replace('"', '\\"')
    cmd = ["mmcli", "-m", mm_index, f'--messaging-create-sms=number={phone_number},text="{escaped}"']
    code, out, err = _run(cmd)
    if code != 0:
        return False, err

    # 从输出 "SMS /org/.../SMS/0 successfully created" 中提取短信对象索引
    match = re.search(r"/SMS/(\d+)", out)
    if not match:
        return False, "Could not find created SMS index"

    sms_idx = match.group(1)
    try:
        code2, out2, err2 = _run(["mmcli", "-m", mm_index, "-s", sms_idx, "--send"], timeout=20)
    except subprocess.TimeoutExpired:
        _run(["mmcli", "-m", mm_index, "-s", sms_idx, "--delete"])
        return False, "发送超时：设备已注册但网络拒绝短信，请确认SIM卡已开通短信服务或VoLTE功能"
    if code2 != 0:
        _run(["mmcli", "-m", mm_index, "-s", sms_idx, "--delete"])
        return False, err2

    _run(["mmcli", "-m", mm_index, "-s", sms_idx, "--delete"])
    return True, "sent"


def list_inbox(mm_index: str) -> List[Dict[str, Any]]:
    """List received SMS messages for a modem."""
    code, out, err = _run(["mmcli", "-m", mm_index, "--messaging-list-sms", "-J"])
    if code != 0:
        return []
    try:
        data = json.loads(out)
        paths = data.get("modem.messaging.sms", [])
        messages = []
        for path in paths:
            match = re.search(r"/SMS/(\d+)$", path)
            if not match:
                continue
            sms_idx = match.group(1)
            code2, out2, _ = _run(["mmcli", "-m", mm_index, "-s", sms_idx, "-J"])
            if code2 == 0:
                try:
                    sms_data = json.loads(out2)
                    sms = sms_data.get("sms", {}).get("content", {})
                    props = sms_data.get("sms", {}).get("properties", {})
                    # 只收录真正收到的短信（deliver），排除发出/待发的（submit）
                    if props.get("pdu-type", "") != "deliver":
                        continue
                    messages.append({
                        "sms_index": sms_idx,
                        "phone_number": sms.get("number", ""),
                        "content": sms.get("text", ""),
                        "timestamp": props.get("timestamp", ""),
                        "state": props.get("state", ""),
                    })
                except Exception:
                    pass
        return messages
    except Exception:
        return []


def delete_sms_from_modem(mm_object_path: str, mm_sms_index: str) -> bool:
    """Delete an SMS object from the modem (inbound cleanup)."""
    if mm_object_path.startswith("zte:"):
        return True  # ZTE SMS index不对应mmcli对象，跳过
    # mm_object_path 形如 /org/freedesktop/ModemManager1/Modem/0
    # SMS D-Bus 路径形如 /org/freedesktop/ModemManager1/SMS/<index>
    base = mm_object_path.rsplit("/Modem/", 1)[0]
    sms_path = f"{base}/SMS/{mm_sms_index}"
    match = re.search(r"/Modem/(\d+)$", mm_object_path)
    if not match:
        return False
    mm_index = match.group(1)
    code, _, _ = _run(["mmcli", "-m", mm_index, f"--messaging-delete-sms={sms_path}"])
    return code == 0


def enable_modem(mm_index: str) -> bool:
    """Enable a disabled modem so it can register to the network."""
    code, _out, err = _run(["mmcli", "-m", mm_index, "-e"])
    if code != 0:
        logger.error(f"mmcli enable modem {mm_index} failed: {err}")
    return code == 0

File: backend-go/services/modem_manager.go
package services

import (
	"encoding/json"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// ModemInfo mirrors the dict returned by the Python modem_manager.
type ModemInfo struct {
	MmObjectPath       string
	MmIndex            string
	DevicePath         string
	Manufacturer       string
	Model              string
	Imei               string
	Operator           string
	SignalQuality      int
	RawState           string
	Status             string
	PhoneNumber        string
	AccessTechnologies string
	RegistrationState  string
	TxBytes            int64
	RxBytes            int64
	ConnectionDuration int64
	Imsi               string
	Iccid              string
	FirmwareRevision   string
	HardwareRevision   string
	CurrentBands       string
	SimOperatorName    string
	SimOperatorCode    string
	CurrentModes       string
	Ports              string
	Plugin             string
	Source             string // "zte" for ZTE devices
}

// InboxMessage is a received SMS parsed from the modem.
type InboxMessage struct {
	SmsIndex    string
	PhoneNumber string
	Content     string
	Timestamp   string
	State       string
}

var (
	reModemIdx  = regexp.MustCompile(`/Modem/(\d+)$`)
	reBearerIdx = regexp.MustCompile(`/Bearer/(\d+)$`)
	reSimIdx    = regexp.MustCompile(`/SIM/(\d+)$`)
	reSmsIdx    = regexp.MustCompile(`/SMS/(\d+)`)
	reSmsIdxEnd = regexp.MustCompile(`/SMS/(\d+)$`)
	reCnum      = regexp.MustCompile(`\+CNUM:.*?"(\+?\d+)"`)
)

var errTimeout = &execErr{"command timed out"}

type execErr struct{ msg string }

func (e *execErr) Error() string { return e.msg }

func run(timeout time.Duration, args ...string) (string, string, error) {
	cmd := exec.Command(args[0], args[1:]...)
	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Start(); err != nil {
		return "", "", err
	}
	done := make(chan error, 1)
	go func() { done <- cmd.Wait() }()
	select {
	case err := <-done:
		return strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String()), err
	case <-time.After(timeout):
		_ = cmd.Process.Kill()
		<-done
		return "", "timeout", errTimeout
	}
}

// ListModems returns all modems detected by ModemManager.
func ListModems() []ModemInfo {
	out, _, err := run(30*time.Second, "mmcli", "-L", "-J")
	if err != nil {
		return nil
	}
	var data struct {
		ModemList []string `json:"modem-list"`
	}
	if json.Unmarshal([]byte(out), &data) != nil {
		return nil
	}
	var modems []ModemInfo
	for _, p := range data.ModemList {
		if info := GetModemInfo(p); info != nil {
			modems = append(modems, *info)
		}
	}
	return modems
}

// GetModemInfo fetches detail for a single modem by D-Bus path.
func GetModemInfo(mmPath string) *ModemInfo {
	m := reModemIdx.FindStringSubmatch(mmPath)
	if m == nil {
		return nil
	}
	idx := m[1]
	out, _, err := run(30*time.Second, "mmcli", "-m", idx, "-J")
	if err != nil {
		return nil
	}
	var raw struct {
		Modem struct {
			Generic map[string]json.RawMessage `json:"generic"`
			Threegp map[string]json.RawMessage `json:"3gpp"`
		} `json:"modem"`
	}
	if json.Unmarshal([]byte(out), &raw) != nil {
		return nil
	}
	g := raw.Modem.Generic
	tg := raw.Modem.Threegp

	sig := 0
	if v, ok := g["signal-quality"]; ok {
		var sq struct {
			Value string `json:"value"`
		}
		if json.Unmarshal(v, &sq) == nil {
			sig, _ = strconv.Atoi(sq.Value)
		}
	}
	state := asString(g["state"])
	bearer := getBearerStats(idx)
	sim := getSimInfo(asString(g["sim"]))
	phone := parseOwnNumber(g)
	if phone == "" {
		phone = getPhoneNumber(idx)
	}

	return &ModemInfo{
		MmObjectPath:       mmPath,
		MmIndex:            idx,
		DevicePath:         asString(g["primary-port"]),
		Manufacturer:       asString(g["manufacturer"]),
		Model:              asString(g["model"]),
		Imei:               asString(tg["imei"]),
		Operator:           asString(tg["operator-name"]),
		SignalQuality:      sig,
		RawState:           state,
		Status:             mapState(state),
		PhoneNumber:        phone,
		AccessTechnologies: joinList(g["access-technologies"]),
		RegistrationState:  asString(tg["registration-state"]),
		TxBytes:            bearer.tx,
		RxBytes:            bearer.rx,
		ConnectionDuration: bearer.dur,
		Imsi:               sim.imsi,
		Iccid:              sim.iccid,
		FirmwareRevision:   asString(g["revision"]),
		HardwareRevision:   asString(g["hardware-revision"]),
		CurrentBands:       joinList(g["current-bands"]),
		SimOperatorName:    sim.opName,
		SimOperatorCode:    sim.opCode,
		CurrentModes:       asString(g["current-modes"]),
		Ports:              joinList(g["ports"]),
		Plugin:             asString(g["plugin"]),
	}
}

type bearerStats struct{ tx, rx, dur int64 }

func getBearerStats(idx string) bearerStats {
	out, _, err := run(30*time.Second, "mmcli", "-m", idx, "--list-bearers", "-J")
	if err != nil {
		return bearerStats{}
	}
	var data struct {
		DBusPath   []string `json:"modem.bearers.dbus-path"`
		BearerList []string `json:"bearer-list"`
	}
	if json.Unmarshal([]byte(out), &data) != nil {
		return bearerStats{}
	}
	paths := data.DBusPath
	if len(paths) == 0 {
		paths = data.BearerList
	}
	if len(paths) == 0 {
		return bearerStats{}
	}
	bm := reBearerIdx.FindStringSubmatch(paths[0])
	if bm == nil {
		return bearerStats{}
	}
	out2, _, err := run(30*time.Second, "mmcli", "-b", bm[1], "-J")
	if err != nil {
		return bearerStats{}
	}
	var b struct {
		Bearer struct {
			Stats  map[string]string `json:"stats"`
			Status map[string]string `json:"status"`
		} `json:"bearer"`
	}
	if json.Unmarshal([]byte(out2), &b) != nil {
		return bearerStats{}
	}
	return bearerStats{
		tx:  atoi64(b.Bearer.Stats["tx-bytes"]),
		rx:  atoi64(b.Bearer.Stats["rx-bytes"]),
		dur: atoi64(b.Bearer.Status["connection-duration"]),
	}
}

type simInfo struct{ imsi, iccid, opName, opCode string }

func getSimInfo(simPath string) simInfo {
	if simPath == "" {
		return simInfo{}
	}
	m := reSimIdx.FindStringSubmatch(simPath)
	if m == nil {
		return simInfo{}
	}
	out, _, err := run(30*time.Second, "mmcli", "-i", m[1], "-J")
	if err != nil {
		return simInfo{}
	}
	var data struct {
		Sim struct {
			Properties map[string]string `json:"properties"`
		} `json:"sim"`
	}
	if json.Unmarshal([]byte(out), &data) != nil {
		return simInfo{}
	}
	p := data.Sim.Properties
	return simInfo{p["imsi"], p["iccid"], p["operator-name"], p["operator-code"]}
}

func parseOwnNumber(g map[string]json.RawMessage) string {
	if v, ok := g["own-numbers"]; ok {
		var arr []string
		if json.Unmarshal(v, &arr) == nil && len(arr) > 0 {
			return arr[0]
		}
	}
	return ""
}

func getPhoneNumber(idx string) string {
	out, _, err := run(30*time.Second, "mmcli", "-m", idx, "--command=AT+CNUM", "-J")
	if err != nil {
		return ""
	}
	var data struct {
		Modem struct {
			Command struct {
				Response string `json:"response"`
			} `json:"command"`
		} `json:"modem"`
	}
	if json.Unmarshal([]byte(out), &data) != nil {
		return ""
	}
	if m := reCnum.FindStringSubmatch(data.Modem.Command.Response); m != nil {
		return m[1]
	}
	return ""
}

func mapState(state string) string {
	switch strings.ToLower(state) {
	case "registered", "connected":
		return "connected"
	case "disabled", "disabling", "enabling", "searching":
		return "disconnected"
	case "failed":
		return "error"
	}
	return "unknown"
}

// SendSMS creates, sends and deletes an SMS object. Returns (success, message).
func SendSMS(mmIndex, phoneNumber, text string) (bool, string) {
	escaped := strings.ReplaceAll(text, `"`, `\"`)
	createArg := "--messaging-create-sms=number=" + phoneNumber + `,text="` + escaped + `"`
	out, stderr, err := run(30*time.Second, "mmcli", "-m", mmIndex, createArg)
	if err != nil {
		return false, stderr
	}
	m := reSmsIdx.FindStringSubmatch(out)
	if m == nil {
		return false, "Could not find created SMS index"
	}
	smsIdx := m[1]
	_, stderr2, err2 := run(20*time.Second, "mmcli", "-m", mmIndex, "-s", smsIdx, "--send")
	if err2 == errTimeout {
		run(30*time.Second, "mmcli", "-m", mmIndex, "-s", smsIdx, "--delete")
		return false, "发送超时：设备已注册但网络拒绝短信，请确认SIM卡已开通短信服务或VoLTE功能"
	}
	if err2 != nil {
		run(30*time.Second, "mmcli", "-m", mmIndex, "-s", smsIdx, "--delete")
		return false, stderr2
	}
	run(30*time.Second, "mmcli", "-m", mmIndex, "-s", smsIdx, "--delete")
	return true, "sent"
}

// ListInbox returns received (deliver) SMS messages for a modem.
func ListInbox(mmIndex string) []InboxMessage {
	out, _, err := run(30*time.Second, "mmcli", "-m", mmIndex, "--messaging-list-sms", "-J")
	if err != nil {
		return nil
	}
	var data struct {
		Paths []string `json:"modem.messaging.sms"`
	}
	if json.Unmarshal([]byte(out), &data) != nil {
		return nil
	}
	var msgs []InboxMessage
	for _, p := range data.Paths {
		m := reSmsIdxEnd.FindStringSubmatch(p)
		if m == nil {
			continue
		}
		smsIdx := m[1]
		out2, _, e := run(30*time.Second, "mmcli", "-m", mmIndex, "-s", smsIdx, "-J")
		if e != nil {
			continue
		}
		var sd struct {
			Sms struct {
				Content    map[string]string `json:"content"`
				Properties map[string]string `json:"properties"`
			} `json:"sms"`
		}
		if json.Unmarshal([]byte(out2), &sd) != nil {
			continue
		}
		if sd.Sms.Properties["pdu-type"] != "deliver" {
			continue
		}
		msgs = append(msgs, InboxMessage{
			SmsIndex:    smsIdx,
			PhoneNumber: sd.Sms.Content["number"],
			Content:     sd.Sms.Content["text"],
			Timestamp:   sd.Sms.Properties["timestamp"],
			State:       sd.Sms.Properties["state"],
		})
	}
	return msgs
}

// DeleteSmsFromModem removes an inbound SMS object from the physical modem.
func DeleteSmsFromModem(mmObjectPath, mmSmsIndex string) bool {
	if strings.HasPrefix(mmObjectPath, "zte:") {
		return true
	}
	base := strings.SplitN(mmObjectPath, "/Modem/", 2)[0]
	smsPath := base + "/SMS/" + mmSmsIndex
	m := reModemIdx.FindStringSubmatch(mmObjectPath)
	if m == nil {
		return false
	}
	_, _, err := run(30*time.Second, "mmcli", "-m", m[1], "--messaging-delete-sms="+smsPath)
	return err == nil
}

// EnableModem enables a disabled modem.
func EnableModem(mmIndex string) bool {
	_, _, err := run(30*time.Second, "mmcli", "-m", mmIndex, "-e")
	return err == nil
}

func asString(r json.RawMessage) string {
	if r == nil {
		return ""
	}
	var s string
	if json.Unmarshal(r, &s) == nil {
		return s
	}
	return strings.Trim(string(r), `"`)
}

func joinList(r json.RawMessage) string {
	if r == nil {
		return ""
	}
	var arr []string
	if json.Unmarshal(r, &arr) == nil {
		return strings.Join(arr, ",")
	}
	return asString(r)
}

func atoi64(s string) int64 {
	n, _ := strconv.ParseInt(s, 10, 64)
	return n
}
